mllib/src/main/scala/org/apache/spark/ml/tree/impl/TrainingInfo.scala

```diff
@@ -0,0 +1,144 @@
+   *   For each (previously) active node,
+   *     Compute bitset indicating whether each training instance under the node splits left/right
+   *     For each column,
+   *       Sort corresponding range of instances based on bitset.
+   * Update nodeOffsets, activeNodes:
+   *   Split offsets for nodes which split (which can be identified using the bitset).
+   *
+   * @return Updated partition info
+   */
+  def update(splits: Array[Array[Split]], newActiveNodes: Array[LearningNode]): TrainingInfo = {
+    // Create buffers for storing our new arrays of node offsets & impurities
+    val newNodeOffsets = new ArrayBuffer[(Int, Int)]()
+    // Update (per-node) sorting of each column to account for creation of new nodes
+    var nodeIdx = 0
+    while (nodeIdx < activeNodes.length) {
+      val node = activeNodes(nodeIdx)
+      // Get new active node offsets from active nodes that were split
+      if (node.split.isDefined) {
+        // Get split and FeatureVector corresponding to feature for split
+        val split = node.split.get
+        val col = columns(split.featureIndex)
+        val (from, to) = nodeOffsets(nodeIdx)
+        // Compute bitset indicating whether each training example splits left/right
+        val bitset = TrainingInfo.bitSetFromSplit(col, from, to, split, splits)
+        // Update each column according to the bitset
+        val numRows = to - from
+        // Allocate shared temp buffers (shared across all columns) for reordering
+        // feature values/indices for current node.
+        val tempVals = new Array[Int](numRows)
+        val tempIndices = new Array[Int](numRows)
+        val numLeftRows = numRows - bitset.cardinality()
+        columns.foreach(_.updateForSplit(from, to, numLeftRows, tempVals, tempIndices, bitset))
+
+        // Add new node offsets to array
+        val leftIndices = (from, from + numLeftRows)
+        val rightIndices = (from + numLeftRows, to)
+        newNodeOffsets ++= Array(leftIndices, rightIndices)
+      }
+      nodeIdx += 1
+    }
+    TrainingInfo(columns, instanceWeights, newNodeOffsets.toArray, newActiveNodes)
+  }
+
+}
+
+/** Training-info specific utility methods. */
+private[impl] object TrainingInfo {
+  /**
+   * For a given feature, for a given node, apply a split and return a bitset indicating the
+   * outcome of the split for each instance at that node.
+   *
+   * @param col  Column for feature
+   * @param from  Start offset in col for the node
+   * @param to  End offset in col for the node
+   * @param split  Split to apply to instances at this node.
+   * @return  Bitset indicating splits for instances at this node.
+   *          These bits are sorted by the row indices.
+   *          bitset(i) = true if ith example for current node splits right, false otherwise.
+   */
+  private[impl] def bitSetFromSplit(
+      col: FeatureVector,
+      from: Int,
+      to: Int,
+      split: Split,
+      allSplits: Array[Array[Split]]): BitSet = {
```

Comment:
Ditto, you only need to pass in the featureSplit: Array[Split], don't pass all splits for all features.